appcelerator/titanium_mobile
# feat(android): fire touch events synchronously

**JIRA:**
https://jira.appcelerator.org/browse/TIMOB-28286

**Summary:**
Modified `touchstart`, `touchmove`, `touchend`, and `touchcancel` events to fire immediately/synchronously like iOS. Needed to implement drag-and-drop interface so it won't look jumpy/laggy.

---
**Test 1:**
1. Build and run the below on Android.
2. Drag one of the squares onscreen.
3. Verify the square drags smoothly under your finger.

```javascript
function createDragSquare(top, left, color) {
	var view = Ti.UI.createView({
		touchEnabled: true,
		backgroundColor: color,
		center: { x: left + 50, y: top + 50 },
		width: 100,
		height: 100,
	});
	view.addEventListener("touchmove", function(event) {
		view.center = view.convertPointToView({ x: event.x, y: event.y }, window);
		event.cancelBubble = true;
	});
	return view;
}

var window = Ti.UI.createWindow({ backgroundColor: "white" });
window.add(createDragSquare(40, 40, "red"));
window.add(createDragSquare(120, 120, "green"));
window.add(createDragSquare(200, 200, "blue"));
window.open();
```

---
**Test 2:**
_(Verifies touch events bubbled to parent.)_
1. Build and run the below on Android.
2. Drag the blue square onscreen
3. Verify the blue square drags smoothly under your finger.
4. Drag your finger on the white background.
5. Verify a green square drags smoothly under you finger.

```javascript
const window = Ti.UI.createWindow({ backgroundColor: "white" });
const view1 = Ti.UI.createView({
	backgroundColor: "blue",
	width: 150,
	height: 150,
});
window.add(view1);
const view2 = Ti.UI.createView({
	visible: false,
	backgroundColor: "green",
	width: 150,
	height: 150,
});
window.add(view2);
function onTouch(e) {
	if (e.source == view1) {
		view1.center = view1.convertPointToView({ x: e.x, y: e.y }, window);
	} else if (e.source == window) {
		switch (e.type) {
			case "touchstart":
				view2.visible = true;
			case "touchmove":
				view2.center = { x: e.x, y: e.y };
				break;
			default:
				view2.visible = false;
		}
	}
}
window.add(Ti.UI.createLabel({
	text: "Drag Square",
	bottom: "30dp",
	color: "black",
	touchEnabled: false,
}));
window.addEventListener("touchstart", onTouch);
window.addEventListener("touchmove", onTouch);
window.addEventListener("touchend", onTouch);
window.addEventListener("touchcancel", onTouch);
window.open();
```
